### PR TITLE
fix #2677

### LINKED ISSUE #2677: allow metagenomic - per sample fastq ebi submissions
AFAIK this is possible but only via CLI, we should open via GUI

### PR DESCRIPTION
**Pull request by antgonza**
Now the forward read will be named sample_name.R1.fastq.gz (before it was sample_name.fastq.gz); the reverse will be R2. There are not a lot of changes, just lots of new tests.

### CODE REVIEW

File: qiita_ware/test/test_ebi.py [hunk 1]
```diff
@@ -1480,8 +1610,8 @@ def test_parse_EBI_reply(self):
     <EXPERIMENT_REF accession="ERX0000008" />
     <DATA_BLOCK>
       <FILES>
-        <FILE checksum="938c29679790b9c17e4dab060fa4c8c5" \
-checksum_method="MD5" filename="%(ebi_dir)s/1.SKB2.640194.fastq.gz" \
+        <FILE checksum="a32357beb845f5b598f1a712fb3b4c70" \
```
[the remark this answers is not in the document]
Comment (antgonza) [reply qiita_ware/test/test_ebi.py]: AFAIK due to the gzip and the timestamp of creation.

File: qiita_pet/handlers/study_handlers/ebi_handlers.py [hunk 2]
```diff
@@ -53,27 +56,39 @@ def display_template(self, preprocessed_data_id, msg, msg_level):
...
+                msg_level = 'success'
+        elif artifact_type == 'per_sample_FASTQ':
+            raw_forward_seqs = []
+            raw_reverse_seqs = []
+            for _, path, ftype in artifact.filepaths:
+                if ftype == 'raw_forward_seqs':
+                    raw_forward_seqs.append(path)
+                elif ftype == 'raw_reverse_seqs':
+                    raw_reverse_seqs.append(path)
+            stats.append(('Total forward', len(raw_forward_seqs)))
```
[the remark this answers is not in the document]
Comment (wasade) [reply qiita_pet/handlers/study_handlers/ebi_handlers.py]: What about the "Number of Sequences" added on line 80 that is conditionally added, and is not added if `artifact_type == 'per_sample_FASTQ'`? Similarly, it looks like total forward and total reverse is branch specific

File: qiita_ware/ebi.py [hunk 3]
```diff
@@ -587,6 +595,20 @@ def generate_run_xml(self):
                 'checksum': md5}
             )
 
+            # inserting reverse
+            if self.per_sample_FASTQ_reverse:
+                file_path = self.sample_demux_fps[
```
Comment (wasade) [inline qiita_ware/ebi.py]: I think the next 10 lines of code are identical to the 10 lines starting on line 586 with the exception of the suffix used? What about

```python
def _add_file_subelement(self, add_file, sample_name, is_forward):
    if is_forward:
        suffix = self.FWD_READ_SUFFIX
    else:
        suffix = self.REV_READ_SUFFIX

    file_path = self.sample_demux_fps[sample_name] + suffix
    with open(file_path) as fp:
        md5 = safe_md5(fp).hexdigest()
    
    file_details = {'filetype': file_type,
                    'quality_scoring_system': 'phred',
                    'checksum_method': 'MD5',
                    'checksum': md5,
                    'filename': join(self.ebi_dir, basename(file_path))}
   
    add_file(file_details)
```

...and then in `generate_run_xml`, do:

```python
    add_file = functools.partial(ET.SubElement, files, 'FILE')
    self._add_file_subelement(add_file, sample_name, is_forward=False)
    if self.per_sample_FASTQ_reverse:
        _add_file_subelement(add_file, sample_name, is_forward=False)
```

Testing of `_add_file_subelement` can be done with a mocked out ET as well so no need to spin up complex ET logic.

File: qiita_ware/ebi.py [hunk 4]
```diff
@@ -931,32 +956,54 @@ def data_retriever(key, trans_dict):
...
-        fps = [(basename(fp), fp) for _, fp, fpt in ar.filepaths
-               if fpt == 'raw_forward_seqs']
-        fps.sort(key=lambda x: x[1])
+
+        # helper function to write files in this method
+        def _rename_file(fp, new_fp):
+            if fp.endswith('.gz'):
+                copyfile(fp, new_fp)
+            else:
+                cmd = "gzip -c %s > %s" % (fp, new_fp)
```
[the remark this answers is not in the document]
Comment (wasade) [reply qiita_ware/ebi.py]: Thanks!

File: qiita_ware/ebi.py [hunk 5]
```diff
@@ -1053,13 +1110,28 @@ def generate_demultiplexed_fastq(self, rewrite_fastq=False, mtime=None):
...
                 demux_samples = self._generate_demultiplexed_fastq_demux(mtime)
         else:
+            # if we are within this else, it means that we already have
+            # generated the raw files and for some reason the submission
+            # failed so we don't need to generate the files again and just
+            # check which files exist in the file path to create our final
+            # list of samples
             demux_samples = set()
-            extension = '.fastq.gz'
+            extension = '.R1.fastq.gz'
```
Comment (wasade) [inline qiita_ware/ebi.py]: should this be `self.FWD_READ_SUFFIX`?

File: qiita_ware/test/test_ebi.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (wasade) [reply qiita_ware/test/test_ebi.py]: fastq doesn't have a timestamp though, so not sure I follow. This suggests the sequences, the sequence identifiers, the phred scores, or the order of the records changed in this sequence file.

File: qiita_pet/handlers/study_handlers/ebi_handlers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (antgonza) [reply qiita_pet/handlers/study_handlers/ebi_handlers.py]: That is correct. Now, not sure what the suggestion is, could you clarify? Note that per_sample_FASTQ doesn't have an easy way to collect the number of sequences.

File: qiita_ware/ebi.py (hunk 3 above)
Comment (antgonza) [reply qiita_ware/ebi.py]: Like it! ... but what about making _add_file_subelement a function that only exists in generate_run_xml? Which is the only place where is being used ...

File: qiita_ware/ebi.py (hunk 5 above)
Comment (antgonza) [reply qiita_ware/ebi.py]: yup, will fix.

**Comment by antgonza** [comment]
Note that added some `self.files_to_remove.append(submission.full_ebi_dir)` in the test as I noticed that some of the tests were not cleaning after them and meant that we were testing with already generated data. It seems like nothing was affected but better safe than sorry.

**Comment by wasade** [comment]
Good to merge?

**Comment by antgonza** [comment]
Yes, please!

**Pull request merged**